src/routes/snippets.rs

```diff
@@ -0,0 +1,82 @@
+use std::collections::BTreeSet;
+use std::path::PathBuf;
+
+use rocket::http::uri::Origin;
+use rocket::response::status::Created;
+use rocket::State;
+use rocket_contrib::json::JsonValue;
+use serde::Deserialize;
+
+use crate::application::Config;
+use crate::errors::ApiError;
+use crate::storage::{Changeset, Snippet, Storage};
+use crate::util::Input;
+
+#[derive(Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct NewSnippet {
+    /// Snippet title. May be omitted in the user request.
+    pub title: Option<String>,
+    /// Snippet syntax. May be omitted in the user request.
+    pub syntax: Option<String>,
+    /// Snippet content. Must be specified in the user request. Can't be empty.
+    pub content: String,
+    /// List of tags attached to the snippet. May be omitted in the user
+    /// request.
+    pub tags: Option<Vec<String>>,
+}
+
+impl NewSnippet {
+    pub fn validate(
+        self,
+        allowed_syntaxes: Option<&BTreeSet<String>>,
+    ) -> Result<Snippet, ApiError> {
+        if self.content.is_empty() {
+            return Err(ApiError::BadRequest(String::from(
+                "`content` - empty values not allowed.",
+            )));
+        }
+        if let Some(syntax) = &self.syntax {
+            if let Some(allowed_syntaxes) = &allowed_syntaxes {
+                if !allowed_syntaxes.contains(syntax) {
+                    return Err(ApiError::BadRequest(format!(
+                        "`syntax` - unallowed value {}.",
+                        syntax
+                    )));
+                }
+            }
+        }
+
+        Ok(Snippet::new(
+            self.title,
+            self.syntax,
+            vec![Changeset::new(0, self.content)],
+            self.tags.unwrap_or_default(),
+        ))
+    }
+}
+
+fn build_location(origin: &Origin, relative_path: &str) -> Result<String, ApiError> {
+    let new_path = PathBuf::from(origin.path()).join(relative_path);
```

Comment:
nit: I don't think the produces string will have `/` slashes on Windows, and URLs expect `/` slashes. We will probably need to use `https://docs.rs/url/2.2.0/url/` in the future. I'm ok to stick with it for a while and check it later with tests on Windows.

Reply:
Good point! Let me add a TODO about that.